packetbeat/protos/dns/dns.go

```diff
@@ -707,3 +721,170 @@ func decodeDnsPacket(data []byte) (dns *layers.DNS, err error) {
 	}
 	return d, nil
 }
+
+// TCP implementation
+
+func (dns *Dns) Parse(pkt *protos.Packet, tcpTuple *common.TcpTuple, dir uint8, private protos.ProtocolData) protos.ProtocolData {
+	defer logp.Recover("DNS ParseTcp")
+
+	logp.Debug("dns", "Parsing packet addressed with %s of length %d.",
+		pkt.Tuple.String(), len(pkt.Payload))
+
+	priv := dnsPrivateData{}
+
+	if private != nil {
+		var ok bool
+		priv, ok = private.(dnsPrivateData)
+		if !ok {
+			priv = dnsPrivateData{}
+		}
+	}
+
+	var payload []byte
+
+	// Offset is critical
```

Comment:
Here is the reason for the two byte offset. "The message is prefixed with a two byte length field which gives the message length, excluding the two byte length field." - [RFC 1035](https://www.ietf.org/rfc/rfc1035.txt) - 4.2.2

This code should use the length field to short-circuit any processing. There is no need to attempt to decode the data when number of bytes is less than the length.